articles/130_ros_time.md

```diff
@@ -115,23 +127,38 @@ It is possible that the user may have access to an out of band time source which
 It might be possible that for their use case a more advanced algorithm would be needed to propagate the simulated time with adequate precision or latency with restricted bandwidth or connectivity.
 The user will be able to switch out the time source for either each instance of their Time object as well as have the ability to override the default for the process.
 
-It is possible to use an external time source such as GPS in as a ROSTime source, but it is recommended to integrate a time source like that using standard ntp integrations with the system clock since that is already an established mechanism and will not need to deal with more complicated changes such as time jumps.
+It is possible to use an external time source such as GPS in as a ROSTime source, but it is recommended to integrate a time source like that using standard NTP integrations with the system clock since that is already an established mechanism and will not need to deal with more complicated changes such as time jumps.
+
+For the current implementation a `TimeSource` API will be defined such that it can be overridden in code.
+If in the future a common implementation is found that would be generally useful it could be extended to optionally dynamically select the alternative TimeSource via a parameter similar to enabling the simulated time.
 
 ## Implementation
 
 The `SystemTime`, `SteadyTime`, and `ROSTime` API's will be provided by each client library in an idiomatic way, but they may share a common implementation, e.g. provided by `rcl`.
 However, if a client library chooses to not use the shared implementation then it must implement the functionality itself.
 
+`SteadyTime` will be typed differently than the interchangable `SystemTime` and `ROSTime`.
+
+
 ### Public API
 
-In each implementation will provide `Time`, `Duration`, and `Rate` datatypes, for all three time source abstraction..
-The `Duration` will support a `sleep_for` function as well as a `sleep_until` method.
+In each implementation will provide `Time`, `Duration`, and `Rate` datatypes, for all three time source abstractions.
+
+The `Clock` will support a `sleep_for` function as well as a `sleep_until` method using a `Duration` or `Time` argument.
 The implementation will also provide a `Timer` object which will provide periodic callback functionality for all the abstractions.
 
+It will also support registering callback before and after a time jump.
+The first callback will be to allow proper preparations for a time jump.
+The latter will allow code to respond to the change in time and include the new time specifically as well as a quantification of the jump.
+
+Any API which is blocking will allow a set of flags to indicate the appropriate behavior in case of time jump.
+This will allow the user to choose to error immediately on a time jump or choose to ignore.
```

Comment:
What is considered to be a time jump in this context? When the clock moves backward? Or also when it moves forward "by a lot" - what would that threshold be?

Reply:
I plan to define a struct that will allow you to define what you consider a "jump" when you register the callback. Such that you will only get the callback when a jump exceeds your threshold.

Right now I see 3 dimensions. There's a backwards jump of size N, a forwards jump of size N, and a change in clock type (aka ROSTime is disable or enabled)

I would expect the default to trigger if the clock changes or there's a backwards jump of any size.

Reply:
That sounds reasonable. Can you please add this concept to the article.

Reply:
Also, in the case of sim time, you can know any time the time jumps backwards or irregularly regardless of the size of the jump.